upload_results.py

```diff
@@ -0,0 +1,174 @@
+        raise ValueError(f"Invalid value for argument 'mode': '{args.mode}'.")
+
+    return args
+
+
+def find_most_recent_result_folder() -> pathlib.Path:
+    root_dir = pathlib.Path(__file__).parent
+    config = config_load(root_dir / "resources" / "config.yaml")
+    output_dir = pathlib.Path(config.output_dir or default_dirs.output_dir)
+
+    def dirname_to_datetime(dirname: str) -> datetime:
+        _, timestamp = dirname.rsplit('.', 1)
+        return datetime.strptime(timestamp, "%Y%m%dT%H%M%S")
+
+    run_directories = output_dir.glob("*.*.*.*")
+    _, run_directory = max((dirname_to_datetime(str(d)), d) for d in run_directories)
+    return run_directory
+
+
+def resolve_input_directory(path: Optional[pathlib.Path]) -> pathlib.Path:
+    path = path or find_most_recent_result_folder()
+    path = path.expanduser().absolute()
+    if not path.is_dir():
+        raise ValueError(f"{path} is not a directory.")
+    return path
+
+
+def configure_logging(verbose: bool):
+    formatter = logging.Formatter(
+        '[%(asctime)s] %(levelname)s: %(message)s',
+        '%H:%M:%S',
+    )
+    log_level = logging.DEBUG if verbose else logging.INFO
+
+    console = logging.StreamHandler()
+    console.setFormatter(formatter)
+    console.setLevel(log_level)
+
+    log.addHandler(console)
+
+
+@contextmanager
+def _connect_to_test_server():
+    openml.config.start_using_configuration_for_example()
+    yield
+    openml.config.stop_using_configuration_for_example()
+
+
+def server_for_task(task: pathlib.Path):
+    metadata = _load_task_data(task)
+    if metadata.test_server:
+        server_connection = _connect_to_test_server()
+    else:
+        server_connection = contextlib.nullcontext()
+    return server_connection
+
+
+def upload_task(task_directory: pathlib.Path) -> Optional[OpenMLRun]:
+    task_name = os.path.basename(task_directory)
+    try:
+        with server_for_task(task_directory):
+            log.debug("Starting upload for '%s'." % task_name)
+            run = process_task_folder(task_directory)
+            log.info("%s result stored at %s/r/%d"
+                     % (task_name, openml.config.server[:-11], run.id))
+            return run
+    except Exception as e:
+        message = e.message if hasattr(e, "message") else e.args[0]
+        log.warning("Task %s failed to upload: %s" % (task_name, message))
+        if args.fail_fast:
+            raise
+
+
+def process_results(result_dir: pathlib.Path, mode: str = 'check'):
+    prediction_directory = result_dir / "predictions"
+
+    if args.task is None:
+        tasks_to_process = [d for d in prediction_directory.iterdir() if d.is_dir()]
+    elif (prediction_directory / args.task).is_dir():
+        tasks_to_process = [args.task]
+    else:
+        raise ValueError(f"Task '{args.task}' not in '{prediction_directory}'.")
+
+    for task_name in tasks_to_process:
+        full_task_directory = prediction_directory / task_name
```

Comment:
Do you somewhere check that the data is actually from OpenML?

Reply:
Added an explicit check now - previously it wouldn't and it would simply result in an AttributeError.